### PR TITLE
feat(async-csv): Setup Async CSV downloads for Discover

### PR DESCRIPTION
**Pull request by leeandher**
This PR adds the Async CSV feature for Discover queries. Slight changes, the existing button now says 'Export Page', whilst this new button says 'Export All'. Clicking it will do everything it has been set up to do for Issues-by-Tags (i.e. start a task, email the user, expire the file)

image

**Note** This is still hidden behind the `organizations:data-export` feature flag. Without that present on the organization, only the regular 'Export Page' button will show up.

---

I'm not too comfortable merging this just yet, I more over just wanted to put it up so people could see how I've done the implementation.

I'm not sure if this will only handle one 'type' of query, since I took the logic from `src/sentry/api/endpoints/organization_events.py`, and I don't think that's the _only_ discover endpoint so please let me know if I'm on the right track!

cc @bowencai8 

---

**To do:**
- [x] Add tests for async task
- [x] Add a test case for endpoint without Discover feature-flag
- [x] Ensure `issue` field is translated to shortId
- [x] Ensure specified projects are taken from the global header
- [x] Add compatibility for `count()` fields and other modifiers

### CODE REVIEW

**Comment by bowencai8** [comment]
@wmak you might also be interested in taking a look at this since you did some work on the current version.

File: src/sentry/tasks/data_export.py
Comment (leeandher) [inline src/sentry/tasks/data_export.py]: This is probably the important part. I've taken the logic that's in this file, and I'm repeating it in order to generate CSVs without a hard-limit of 50 or 100 rows.

 In terms of infrastructure, this task has its own queue with a concurrency of 1, so no worries there.

I'm just not sure if this code would work for _every_ type of discover query, since I'm not too knowledgeable on how complicated those queries can get. If it works, great! If i missed out on something big (like if you add `x` filter, with `y` range, let me know because I couldn't figure out too much)

**Comment by leeandher** [comment]
Also if you want to test this out locally since you have some cool discover queries set up, just pull it and make sure `organizations:data-export` is set to `True` in `src/sentry/conf/server.py`

File: src/sentry/static/sentry/app/components/dataExport.tsx
Comment (markstory) [inline src/sentry/static/sentry/app/components/dataExport.tsx]: Can we use the Button component here?

File: src/sentry/static/sentry/app/components/gridEditable/index.tsx
Comment (markstory) [inline src/sentry/static/sentry/app/components/gridEditable/index.tsx]: Could use the new download icon in `app/icons/iconDownload`.

File: src/sentry/static/sentry/app/components/gridEditable/index.tsx
Comment (markstory) [inline src/sentry/static/sentry/app/components/gridEditable/index.tsx]: I'm not sure we should have two download buttons beside each other. The async download is what people have been asking for. Getting just the current page was a stop-gap solution.

File: src/sentry/tasks/data_export.py
Comment (markstory) [inline src/sentry/tasks/data_export.py]: Discover queries can often target more than a single project. Will exports only work for one project?

File: src/sentry/tasks/data_export.py
Comment (markstory) [inline src/sentry/tasks/data_export.py]:
```suggestion
            referrer="export.discover",
```
Having a unique referrer will help the snuba team locate problematic query sources. If this shares a name with the API endpoint that gets harder.

File: src/sentry/tasks/data_export.py
Comment (markstory) [inline src/sentry/tasks/data_export.py]: You could hit a KeyError here as results don't always have the `data` key when snuba is overloaded or broken.

File: src/sentry/tasks/data_export.py
Comment (markstory) [inline src/sentry/tasks/data_export.py]:
```suggestion
    # Iterate through all the discover results.
```

File: src/sentry/static/sentry/app/components/gridEditable/index.tsx
Comment (wmak) [reply src/sentry/static/sentry/app/components/gridEditable/index.tsx]: One reason to keep the old download might be if we only have 1 page of results, might be worth swapping this button depending on whether there's a new page or not.

Comment (leeandher) [reply src/sentry/static/sentry/app/components/gridEditable/index.tsx]: @wmak That is probably a good call, I'll change the function names and try to implement that

File: src/sentry/tasks/data_export.py
Comment (leeandher) [reply src/sentry/tasks/data_export.py]: Thanks! these are the type of things i forget to check, im working on a better solution now

File: src/sentry/tasks/data_export.py
Comment (leeandher) [reply src/sentry/tasks/data_export.py]: @markstory If snuba is overloaded or broken, won't it throw an error that'll be managed by the `with snuba_error_handler()`? (i.e. SnubaError), or will the KeyError override that?

Comment (markstory) [reply src/sentry/tasks/data_export.py]: Yes it would. I missed the context manager. My bad.

File: src/sentry/static/sentry/app/components/gridEditable/index.tsx
Comment (markstory) [reply src/sentry/static/sentry/app/components/gridEditable/index.tsx]: Doing a in-place export for single page results is a great idea. If there is more than one page we should offer the async export option.

File: src/sentry/tasks/data_export.py
Comment (leeandher) [reply src/sentry/tasks/data_export.py]: Multi-project queries are supported in the latest changes!

File: src/sentry/static/sentry/app/components/gridEditable/index.tsx
Comment (leeandher) [reply src/sentry/static/sentry/app/components/gridEditable/index.tsx]: @wmak @markstory Made the change!

File: src/sentry/static/sentry/app/components/gridEditable/index.tsx
Comment (leeandher) [reply src/sentry/static/sentry/app/components/gridEditable/index.tsx]: 👌

File: src/sentry/static/sentry/app/components/dataExport.tsx
Comment (leeandher) [reply src/sentry/static/sentry/app/components/dataExport.tsx]: 👌

**Comment by leeandher** [comment]
A big refactor took place between this update and when this PR was opened, but that's all taken care of now.

 - Discover logic is in a `processor` which is just to abstract it from the task
 - Multi-project queries are supported
 - In-browser export is used if results are < 50 (hard-coded for now)
 - Async Export exactly matches column order/layout from the table (but rows may be rearranged)
 - A new function is added to ensure the csv is ascii encoded (can be removed once the `csv` is upgraded to python3)
 - Added an aliasing function to coherce `user` field into `user.name | user.email | user.username | user.ip`

**Please throw some complex queries at it to make sure it works for everything!**

To run it locally make sure `SENTRY_FEATURES["organizations:data-export"]` is `True`, and to force async exports on results < 50, modify the `renderDownloadButton` function in `gridEditable/index.tsx`

File: src/sentry/data_export/tasks.py
Comment (leeandher) [inline src/sentry/data_export/tasks.py]: This will make sure the discover fields in the browser match the export (no extra `user.id` or `user.username` fields)

File: src/sentry/data_export/tasks.py
Comment (leeandher) [inline src/sentry/data_export/tasks.py]: The` __init__` function for `processors/discover.py` does most of the setup.

File: src/sentry/data_export/processors/discover.py
Comment (leeandher) [inline src/sentry/data_export/processors/discover.py]: This file is basically all the actual copied discover code, it's all handled in the constructor, then in the task, we run the `data_fn`.

**Review by leeandher — COMMENTED** [review]
Left a few comments, but hit me up with any questions!

**Comment by leeandher** [comment]
This PR has a lot of extra fluff stuff that could make reviewing it more difficult so I will try to get those into a separate PR and remove them from this diff.

**Comment by leeandher** [comment]
> This PR has a lot of extra fluff stuff that could make reviewing it more difficult so I will try to get those into a separate PR and remove them from this diff.

Grouped a bunch of the small changes into here https://github.com/getsentry/sentry/pull/18024, so this will only be merged afterwards, it'll clear some of the clutter out.

File: src/sentry/data_export/processors/discover.py
Comment (markstory) [inline src/sentry/data_export/processors/discover.py:14]: You might need something similar for `issue` which is another special discover field alias.

File: src/sentry/data_export/processors/discover.py
Comment (markstory) [inline src/sentry/data_export/processors/discover.py:62]: If this function was exposed by the `discover` module you might not need to redefine the alias fields again.

File: src/sentry/static/sentry/app/components/gridEditable/index.tsx
Comment (markstory) [reply src/sentry/static/sentry/app/components/gridEditable/index.tsx:329]: You could use the `<Feature>` component like we do in `renderHeaderButtons()` to restrict access to offline exports with the current single page export being the 'disabled' state for the feature.

File: src/sentry/data_export/processors/discover.py
Comment (leeandher) [reply src/sentry/data_export/processors/discover.py:62]: Are you talking about using a similar function that might be in `src/sentry/snuba/discover.py`?

Comment (leeandher) [reply src/sentry/data_export/processors/discover.py:62]: Are you talking about using a similar function that might be in `src/sentry/snuba/discover.py`?

File: src/sentry/data_export/processors/discover.py
Comment (leeandher) [reply src/sentry/data_export/processors/discover.py:14]: I can't seem to find anywhere to convert the `issue.id` to that unique `shortId` text

Comment (leeandher) [reply src/sentry/data_export/processors/discover.py]: Where would that be?

File: src/sentry/static/sentry/app/components/gridEditable/index.tsx
Comment (leeandher) [reply src/sentry/static/sentry/app/components/gridEditable/index.tsx]: Done!

File: src/sentry/data_export/processors/discover.py
Comment (markstory) [reply src/sentry/data_export/processors/discover.py]: That logic sadly ended up in one of the endpoints. I'd like to clean that up but haven't found the time to do so. https://github.com/getsentry/sentry/blob/df0da0722d626670ccc75b0ae655cfd38e47f4d5/src/sentry/api/bases/organization_events.py#L118-L128

File: src/sentry/data_export/processors/discover.py
Comment (markstory) [reply src/sentry/data_export/processors/discover.py:66]: Yes. That module already has access to the field alias composition and would be a good place to do the coalesce. We'd still need all the component fields in API responses though as the UI needs to do its own coalesce :cry:

Comment (leeandher) [reply src/sentry/data_export/processors/discover.py:66]: I've been picking at this a few days, and I don't think I can directly use the aliasing from (`FIELD_ALIASES`) in `event_search.py` since I need the exported CSV to match what is shown in the browser

Comment (leeandher) [reply src/sentry/data_export/processors/discover.py:66]: I'm going to take a look at replicating some of the stuff in `fieldRenders.tsx` instead, but let me know if this is the wrong approach

File: src/sentry/data_export/processors/discover.py
Comment (leeandher) [reply src/sentry/data_export/processors/discover.py]: @markstory I copied the code over from there, but I know there's a comment to move it to the actual `discover.query` method. Should I do that, or is copying it okay for now?

Comment (markstory) [reply src/sentry/data_export/processors/discover.py]: Copying is fine for now. If you could leave a comment on both letting future us know where its sibling lives that would be great.

File: src/sentry/data_export/processors/discover.py
Comment (markstory) [reply src/sentry/data_export/processors/discover.py]: Reproducing the fieldRenderer code is a good plan.

**Review by leeandher — COMMENTED** [review]
Annotated for round 3 of reviews!

cc @markstory, @wmak 🌮 🌮 

Thanks for helping with these reviews, I know it's a lot to look at

File: src/sentry/data_export/processors/discover.py
Comment (leeandher) [reply src/sentry/data_export/processors/discover.py]: Done!

**Comment by leeandher** [comment]
Removed the user field coalescing since this shouldn't be an issue if merged before https://github.com/getsentry/sentry/pull/18165/ 😄

File: src/sentry/data_export/processors/discover.py
Comment (evanh) [inline src/sentry/data_export/processors/discover.py:34]: This function in the endpoints also handles permissions (does this user have access to this project). Does that check still happen somewhere?

Comment (leeandher) [reply src/sentry/data_export/processors/discover.py:34]: Unfortunately I haven't moved that logic over yet since it's all in the endpoints and it'd be a lot of copying. Ideally I'd want some `util` that both endpoints/my-task can reach into, but I'm not sure if I should be setting that up.

The only way users would be able to get data on projects they shouldn't have access to would be to curl/Postman the request themselves though, it is checked for browser queries

Comment (leeandher) [reply src/sentry/data_export/processors/discover.py:34]: I could start working on that if it's blocking the merge, but I have a separate ticket to address this Jira 👌

File: src/sentry/data_export/processors/discover.py
Comment (markstory) [inline src/sentry/data_export/processors/discover.py:35]: If you normalized `query.project` to a list before starting the export you could have simpler code here.

File: src/sentry/data_export/processors/discover.py
Comment (markstory) [inline src/sentry/data_export/processors/discover.py:53]:
```suggestion
                referrer="task.export.discover",
```
We should have a distinct referrer from the API endpoint so we can more easily discern the performance of each.

File: src/sentry/data_export/tasks.py
Comment (markstory) [inline src/sentry/data_export/tasks.py:130]: This will be sampled by default so you might not get any results. Given that this should be low frequency you might want to set the sampling rate to 1.0.

File: src/sentry/static/sentry/app/components/gridEditable/index.tsx
Comment (markstory) [inline src/sentry/static/sentry/app/components/gridEditable/index.tsx:333]: This looks great. Nice job :clap:

File: src/sentry/data_export/processors/discover.py
Comment (leeandher) [reply src/sentry/data_export/processors/discover.py]: Done!

File: src/sentry/data_export/tasks.py
Comment (leeandher) [reply src/sentry/data_export/tasks.py]: Done!

File: src/sentry/data_export/processors/discover.py
Comment (leeandher) [reply src/sentry/data_export/processors/discover.py:35]: Done!

**Comment by leeandher** [comment]
Euuugh big oversight I just noticed, it doesn't process the computed fields in the export, they show up blank (e.g. `count()` `lastSeen()`). Adding the WIP tag back :((

File: tests/sentry/data_export/endpoints/test_data_export.py
Comment (wedamija) [inline tests/sentry/data_export/endpoints/test_data_export.py:20]: Worth testing project perms here as well.

**Review by wedamija — CHANGES_REQUESTED** [review]

**Comment by leeandher** [comment]
Working on a project permissions fix in a separate PR, will update soon!

File: src/sentry/data_export/endpoints/data_export.py
Comment (leeandher) [reply src/sentry/data_export/endpoints/data_export.py:56]: Will be addressed in https://github.com/getsentry/sentry/pull/18229, It should be merged before this PR to copy over the changes

File: tests/sentry/data_export/endpoints/test_data_export.py
Comment (leeandher) [reply tests/sentry/data_export/endpoints/test_data_export.py:20]: Will add to https://github.com/getsentry/sentry/pull/18229!

File: src/sentry/data_export/processors/discover.py
Comment (leeandher) [reply src/sentry/data_export/processors/discover.py:36]: This will be added in a separate PR -> https://github.com/getsentry/sentry/pull/18229

File: src/sentry/data_export/tasks.py
Comment (fpacifici) [inline src/sentry/data_export/tasks.py:138]: Could you please avoid the `while True` and provide a meaningful condition here, so it is less easy to break this and turning it into an infinite loop?

Comment (leeandher) [reply src/sentry/data_export/tasks.py:138]: I'd actually already used this implementation elsewhere in the file, so I'll try to refactor it in separate ticket for both types of exports `Issues-by-Tag` and these `Discover` ones.

The original reason I'd used a `while True` was because I thought it was easier to read and I have no way of knowing the amount of rows that are going to be in any CSV as they're being assembled.

Comment (leeandher) [reply src/sentry/data_export/tasks.py:138]: Will request your review on the separate ticket though, thanks for the note

**Comment by leeandher** [comment]
Fixed up this PR to:
 - Add a last set of tests
 - Ensure it works with `count(id)`, `last_seen()`, etc. function fields
 - Ensure users see a 403 page/response if they don't have access to the associated projects

**Comment by leeandher** [comment]
Also! Just keep in mind that this will only be visible for the `sentry` and `sentry-test` orgs, as it requires the `organizations:data-export` feature-flag which is currently defaulted to false.

**Comment by markstory** [comment]
Looking good to me. Would be great to have tests for the async tasks that you have in your todo list :smile:

**Comment by leeandher** [comment]
@markstory  My bad forgot to check it off, I tested the two methods in there worth testing!
The tests are in `tests/sentry/data_export/processors/test_discover.py`. 

There aren't that many methods to test like what I had to do for `test_issues_by_tag.py` which is already in the same folder (outside this PR)

File: src/sentry/data_export/tasks.py
Comment (leeandher) [reply src/sentry/data_export/tasks.py]: @fpacifici I overestimated how hard a refactor would be, I added it in this PR let me know what you think!

**Review by evanh — COMMENTED** [review]
All looks good to me

**Review by wedamija — APPROVED** [review]
Looks good to me now that permissions are handled. Probably should also get a +1 from someone from discover before merging.

**Comment by leeandher** [comment]
Gonna merge this in first thing tomorrow, if the lights are still green, thanks everyone for the help!

**Comment by bowencai8** [comment]
Wahoo! I'm excited.

**Review by markstory — APPROVED** [review]

**Pull request merged**
